### ISSUE TITLE
[java] UnusedPrivateField not triggering if @Value annotation present

### ISSUE THREAD

**Issue by OverDrone** [pmd/pmd#1952]
Rule: [UnusedPrivateField](https://pmd.github.io/latest/pmd_rules_java_bestpractices.html#unusedprivatefield)

PMD 6.16 Java.

``` java
class MyClass {
  boolean var;
}
```
correctly produces category\java\bestpractices.xml\UnusedPrivateField warning. BUT:

``` java
@Component
class MyService implements IMyService {
  @Value("...")
  boolean unusedConfigVariable;
}
```

does not produce this warning. I understand that sometimes annotations make private fields usable somehow, but in this case they don't. Can you add a possibility to ignore @Value annotation and produce warning?
Also some previous PMD version, maybe 6.0 produce UnusedPrivateField even if @Value annotation present.

**Comment by jsotuyod** [comment]
@OverDrone this is down to the way you are using PMD. You seem to not be passing along auxclasspath information, so PMD has trouble differentiating between `lombok.Value` and other `@Value` annotations.

When using Gradle, the auxclasspath is passed along automatically.
With the CLI, the  `-auxclasspath` argument should be used. Docs are available [here](https://pmd.github.io/pmd-6.17.0/pmd_userdocs_cli_reference.html#options)
Under Maven, this is done by default, but only under the `verify` goal, if simply running `pmd:pmd` or `pmd:check` the information *may* not be available.
Finally, under Ant, there is an `auxclasspath` node for this, [docs](https://pmd.github.io/pmd-6.17.0/pmd_userdocs_tools_ant.html#parameters) and [examples](https://pmd.github.io/pmd-6.17.0/pmd_userdocs_tools_ant.html#full-example-with-auxclasspath) are available.

Let us know if you need further assistance.

**Comment by OverDrone** [comment]
I'm using PMD both in gradle and in eclipse plugin. Problem is with eclipse plugin (net.sourceforge.pmd.eclipse.feature.group 4.5.0.v20190630-0922). I don't see any option about auxclasspath, only checkbox called smth like "Enable using Java Project Build Path...". If PMD uses Project Build Path, then spring jar with @Value annotation should be on classpath.
Anyway it's unclear to me why PMD needs some classpath to differentiate between lombok.Value and org.springframework.beans.factory.annotation.Value? If you read source file .java, then there is "import org.springframework.beans.factory.annotation.Value;". If you read .class file I'm sure same kind of information is also available. 
And what would happen if auxclasspath would contain both lombok.Value and org.springframework.beans.factory.annotation.Value?

**Comment by adangel** [comment]
So obviously *something* is not working as it should.

> only checkbox called smth like "Enable using Java Project Build Path..."

That's the correct option. This will use the project's build path as "auxclasspath".

> Anyway it's unclear to me why PMD needs some classpath to differentiate between lombok.Value 
> and org.springframework.beans.factory.annotation.Value? If you read source file .java, then there 
> is "import org.springframework.beans.factory.annotation.Value;".

In that simple case, we might be able to find the problem without relying on the classpath. But there are other cases, where we need it, e.g. looking the the inheritance hierarchy of classes/interfaces.

> And what would happen if auxclasspath would contain both lombok.Value and 
> org.springframework.beans.factory.annotation.Value?

It will work. Only if the Spring annotation is not on the classpath, then the following happens:

We don't know the class of `@Value` and we fall back to compare with the simple name - which is just "Value". And this looks like what you are observing...

**Comment by OverDrone** [comment]
> So obviously _something_ is not working as it should.
Can you give me any suggestions as to where to look? switch on logging maybe?

**Comment by adangel** [comment]
If you could create a small sample project, which shows the problem, that would be great. I've done such things earlier, e.g. https://github.com/adangel/scratchpad/tree/master/pmd-issue-1330 . This helps a lot to reproduce the issue. You can also create a gradle based sample project, as long as you can share it with us.

**Comment by OverDrone** [comment]
Sure!
https://github.com/OverDrone/pmd-issue-1952

**Comment by adangel** [comment]
@OverDrone  Thanks for the excellent sample project.

It turned out, that this happens, if lombok is not on the classpath. So, we have a couple of default annotations, we are ignoring. And while verifying, whether these are present or not, we can't resolve them (in that case `lombok.Value`) and we end up comparing just the simple class name - which is "Value". I've seen, that we correctly resolve the annotation, so your auxclasspath was indeed correct.

So we fall back to this line:
https://github.com/pmd/pmd/blob/46a8d87fb7a76a187e6f959171a67eee9562075c/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/typeresolution/TypeHelper.java#L34

That's indeed a serious bug.
